File: scripts/dashboard/system_metrics.py
"""Cached host and process metrics for the read-only dashboard."""
from __future__ import annotations

import copy
import os
import shutil
import subprocess
import threading
import time
from pathlib import Path
from typing import Callable


def _number(value):
    return value if isinstance(value, (int, float)) and not isinstance(value, bool) else None


class SystemMetrics:
    """Collect inexpensive OS counters outside HTTP request handlers."""

    def __init__(self, disk_path: Path, services: Callable[[], dict], collector_service: str | None = None):
        self.disk_path = Path(disk_path)
        self.services = services
        self._lock = threading.Lock()
        self._cache = self._empty()
        self._previous_cpu: tuple[int, int] | None = None
        self.collector_service = collector_service
        self._collector_checked = 0.
        self._collector = {"pid": None, "state": "unknown"}

    @staticmethod
    def _empty() -> dict:
        return {
            "schemaVersion": 1,
            "asOf": None,
            "cpu": {"percent": None, "cores": os.cpu_count()},
            "load": {"one": None, "five": None, "fifteen": None},
            "memory": {"used_bytes": None, "total_bytes": None, "percent": None},
            "disk": {"used_bytes": None, "total_bytes": None, "free_bytes": None, "percent": None},
            "services": {},
            "journal_backlog": None,
            "event_loop_lag_ms": None,
        }

    @staticmethod
    def _cpu_times() -> tuple[int, int] | None:
        try:
            values = [int(value) for value in Path("/proc/stat").read_text(encoding="ascii").splitlines()[0].split()[1:]]
        except (OSError, ValueError, IndexError):
            return None
        if len(values) < 4:
            return None
        idle = values[3] + (values[4] if len(values) > 4 else 0)
        return sum(values), idle

    @staticmethod
    def _memory() -> dict:
        try:
            values = {}
            for line in Path("/proc/meminfo").read_text(encoding="ascii").splitlines():
                key, raw = line.split(":", 1)
                values[key] = int(raw.strip().split()[0]) * 1024
            total, available = values["MemTotal"], values["MemAvailable"]
            used = max(0, total - available)
            return {"used_bytes": used, "total_bytes": total, "percent": used * 100 / total if total else None}
        except (OSError, ValueError, KeyError):
            return {"used_bytes": None, "total_bytes": None, "percent": None}

    @staticmethod
    def _process(pid, state) -> dict:
        active = state == "active"
        result = {"state": state, "pid": pid if active and isinstance(pid, int) and pid > 0 else None,
                  "rss_bytes": None, "uptime_seconds": None}
        if result["pid"] is None:
            return result
        root = Path(f"/proc/{result['pid']}")
        try:
            status = root.joinpath("status").read_text(encoding="ascii")
            rss = next(line for line in status.splitlines() if line.startswith("VmRSS:"))
            result["rss_bytes"] = int(rss.split()[1]) * 1024
            stat = root.joinpath("stat").read_text(encoding="ascii").rsplit(") ", 1)[1].split()
            started = int(stat[19]) / os.sysconf("SC_CLK_TCK")
            uptime = float(Path("/proc/uptime").read_text(encoding="ascii").split()[0])
            result["uptime_seconds"] = max(0., uptime - started)
        except (OSError, ValueError, StopIteration, IndexError):
            pass
        return result

    def _collector_status(self) -> dict:
        if not self.collector_service or os.name == "nt":
            return self._collector
        if time.monotonic() - self._collector_checked < 10:
            return self._collector
        self._collector_checked = time.monotonic()
        try:
            process = subprocess.run(["systemctl", "show", self.collector_service, "--property=MainPID,ActiveState"],
                                     capture_output=True, text=True, timeout=2)
            if process.returncode != 0:
                raise OSError("service state unavailable")
            fields = dict(line.split("=", 1) for line in process.stdout.splitlines() if "=" in line)
            pid = int(fields.get("MainPID", "0"))
            self._collector = {"pid": pid if pid > 0 else None, "state": fields.get("ActiveState") or "unknown"}
        except (OSError, ValueError, subprocess.SubprocessError):
            self._collector = {"pid": None, "state": "unknown"}
        return self._collector

    def refresh(self) -> dict:
        value = self._empty()
        value["asOf"] = time.time()
        current = self._cpu_times()
        if current is not None and self._previous_cpu is not None:
            total = current[0] - self._previous_cpu[0]
            idle = current[1] - self._previous_cpu[1]
            value["cpu"]["percent"] = max(0., min(100., (total - idle) * 100 / total)) if total > 0 else None
        self._previous_cpu = current
        try:
            one, five, fifteen = os.getloadavg()
            value["load"] = {"one": one, "five": five, "fifteen": fifteen}
        except (AttributeError, OSError):
            pass
        value["memory"] = self._memory()
        try:
            disk = shutil.disk_usage(self.disk_path)
            value["disk"] = {"used_bytes": disk.used, "total_bytes": disk.total, "free_bytes": disk.free,
                             "percent": disk.used * 100 / disk.total if disk.total else None}
        except OSError:
            pass
        try:
            service_state = self.services()
        except Exception:
            service_state = {}
        if self.collector_service:
            service_state["collector"] = self._collector_status()
        value["journal_backlog"] = _number(service_state.pop("journal_backlog", None))
        value["event_loop_lag_ms"] = _number(service_state.pop("event_loop_lag_ms", None))
        value["services"] = {
            str(name): self._process(details.get("pid"), str(details.get("state") or "unknown"))
            for name, details in service_state.items() if isinstance(details, dict)
        }
        with self._lock:
            self._cache = value
        return copy.deepcopy(value)

    def snapshot(self) -> dict:
        with self._lock:
            return copy.deepcopy(self._cache)

    def run(self, stop: threading.Event) -> None:
        while not stop.is_set():
            self.refresh()
            stop.wait(1)

File: web/src/system-metrics.ts
import type { SystemMetrics } from './api/types';
import { esc } from './ui';

const value=(v:number|null|undefined,digits=1)=>typeof v==='number'&&Number.isFinite(v)?v.toFixed(digits):'--';
const bytes=(v:number|null|undefined)=>{
  if(typeof v!=='number'||!Number.isFinite(v)||v<0)return '--';
  const units=['B','KB','MB','GB','TB'];let n=v,i=0;
  while(n>=1024&&i<units.length-1){n/=1024;i++;}
  return `${n.toFixed(i<2?0:1)} ${units[i]}`;
};
const serviceNames:Record<string,string>={dashboard:'控制台',collector:'行情采集',trader:'交易进程',projection:'账本投影'};
const stateNames:Record<string,string>={active:'运行中',stopped:'未运行',inactive:'未运行',failed:'失败',activating:'启动中',deactivating:'停止中',unavailable:'不可用',unknown:'未知'};

export function renderSystemMetrics(data:SystemMetrics|null,error:string|null){
  const root=document.querySelector<HTMLElement>('[data-system-metrics]');if(!root)return;
  if(data?.asOf!==null&&data&&Date.now()/1000-data.asOf>5){data=null;error='服务器状态已过期';}
  const set=(name:string,text:string)=>{const node=root.querySelector(`[data-system-${name}]`);if(node)node.textContent=text;};
  set('cpu',data?`${value(data.cpu.percent)}% · ${data.cpu.cores??'--'} 核`:'--');
  set('memory',data?`${value(data.memory.percent)}% · ${bytes(data.memory.used_bytes)} / ${bytes(data.memory.total_bytes)}`:'--');
  set('disk',data?`${value(data.disk.percent)}% · 可用 ${bytes(data.disk.free_bytes)}`:'--');
  set('load',data?`${value(data.load.one,2)} / ${value(data.load.five,2)} / ${value(data.load.fifteen,2)}`:'--');
  const services=root.querySelector('[data-system-services]');
  if(services)services.innerHTML=data?Object.entries(data.services).map(([name,p])=>`<div class="system-service"><span>${esc(serviceNames[name]||name)}</span><strong class="${p.state==='active'?'good':p.state==='stopped'?'':'warn'}">${esc(stateNames[p.state]||p.state)}</strong><small>PID ${p.pid??'--'} · 内存 ${bytes(p.rss_bytes)} · 运行 ${typeof p.uptime_seconds==='number'?`${Math.floor(p.uptime_seconds)} 秒`:'--'}</small></div>`).join(''):'<span class="muted">进程状态未获取</span>';
  set('state',error||(!data?'服务器状态读取中':data.asOf===null?'等待首个系统采样':`更新 ${new Date(data.asOf*1000).toLocaleTimeString('zh-CN',{hour12:false})} · 投影队列 ${data.journal_backlog??'--'}`));
}
